newrelic/common/streaming_utils.py

```diff
@@ -100,12 +113,30 @@ def __next__(self):
                         self.shutdown()
                     raise StopIteration
 
-                try:
-                    return self.stream_buffer._queue.popleft()
-                except IndexError:
-                    pass
-
-                if not self.stream_closed() and not self.stream_buffer._queue:
+                if self.batching:
+                    stream_buffer_len = len(self.stream_buffer)
+                    if stream_buffer_len > self.MAX_BATCH_SIZE:
+                        # Ensure batch size is never more than 100 to prevent issues with serializing large numbers
+                        # of spans causing their age to exceed 10 seconds. That would cause them to be rejected
+                        # by the trace observer.
+                        batch = [self.stream_buffer._queue.popleft() for _ in range(self.MAX_BATCH_SIZE)]
+                        return SpanBatch(spans=batch)
+                    elif stream_buffer_len:
+                        # For small span batches empty stream buffer into list and clear queue.
```

Comment:
This mentions needing to use a lock. Where is the lock?

Reply:
The very first line of the function is the lock. 

`with self._notify:`